get things to compile on JDK 11

eed3si9n opened the pull request:
This is for testing #639 together with #602 to get the baseline failures.

Fixes #540

xuwei-k committed:
get things to compile on jdk11

jvican committed:
Fix mima error in the build

jvican committed:
Test JDK 11 in Drone

eed3si9n committed:
Skip Scala 2.10 and Scala 2.11 components for JDK 11

Scala 2.10 and Scala 2.11 do not support JDK 11.

typesafe-tools commented:
To validate this pull request against other sbt modules, please visit <A HREF="https://jenkins.scala-sbt.org:8592/job/sbt-validator/parambuild/?sbt=develop&zinc=pull/640/head&io=develop&librarymanagement=develop&util=develop&website=develop&deploy_to_bintray=false&run_sbt_community=false&verbose_dbuild=false&community_verbose=false">this link.</A>

eed3si9n committed:
Adjust JavaCompilerSpec to JDK 11

eed3si9n committed:
Bump benchmark commits to sbt 1

sbt 0.13 can't run on JDK 11 because Scala 2.10 doesn't work on JDK 11.
This change the commit id of the benchmark projects so the source slurping code runs against sbt 1 builds.

eed3si9n committed:
Change the multi-project incremental spec to use Scala 2.12

eed3si9n committed:
Remove the test that includes sun.net.spi package

eed3si9n committed:
image: eed3si9n/scala-docs:1.6-M2

eed3si9n commented:
AdoptOpenJDK fixed some of the issues, but now I am hitting https://bugs.openjdk.java.net/browse/JDK-8219218?

eed3si9n commented:
The Docker image `eed3si9n/scala-docs:1.6-M2` changes are here - https://github.com/scalacenter/docker-scala/pull/4

eed3si9n committed:
Remove unused import

eed3si9n committed:
Migrated deprecated methods in JDK11

Since JDK 9 java.lang.Integer constructor is deprecated in favor of valueOf.

Class#newInstace was deprecated in favor of Class#getDeclaredConstructor(...).newInstance(...).

jvican commented on build.sbt:
```diff
@@ -766,28 +768,34 @@ lazy val zincScripted = (project in internalPath / "zinc-scripted")
-      s"${compilerBridgeTest213.id}/test",
-    )
+    val testCommands =
+      (if (isJava8) List(
```
Nitpick: this would be more readable within curly braces than parenthesis. I had a hard time parsing this change and its likely others will too.

eed3si9n replied:
sounds good.

eed3si9n replied:
Added a new `val` - e07e5441dad034eb1b405f60e9bcf53cd02c877a

jvican commented on internal/zinc-benchmarks/src/test/scala/xsbt/ZincBenchmarkSpec.scala:
```diff
@@ -15,50 +15,50 @@ class ZincBenchmarkSpec extends FunSuite {
     }
   }
 
-  test("Check that setup for small project is successful and zinc compiles") {
```
Can we remove this test instead?

eed3si9n replied:
ok.

eed3si9n replied:
https://github.com/sbt/zinc/pull/640/commits/75b3a4b8f9888d7e28abc834f25db7c84f214f48

jvican commented on internal/zinc-compile-core/src/main/scala/sbt/internal/inc/AnalyzingCompiler.scala:
```diff
@@ -239,7 +239,7 @@ final class AnalyzingCompiler(
   )(argTypes: Class[_]*)(args: AnyRef*): AnyRef = {
     val interfaceClass = getInterfaceClass(interfaceClassName, log)
-    val interface = interfaceClass.newInstance.asInstanceOf[AnyRef]
+    val interface = interfaceClass.getDeclaredConstructor().newInstance().asInstanceOf[AnyRef]
```
Does this have the same semantics in JDK 8?

eed3si9n replied:
Yes. `getDeclaredConstructor().newInstance()` is same as calling `new` with the nullary constructor, which is what `Class#newInstance()` did but it's now deprecated.  https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/lang/Class.html#newInstance()

jvican commented on internal/zinc-compile-core/src/test/scala/sbt/internal/inc/javac/JavaCompilerSpec.scala:
```diff
@@ -60,7 +66,9 @@ class JavaCompilerSpec extends UnitSpec with DiagrammedAssertions {
     val classfile = new File(out, "good.class")
     assert(classfile.exists)
-    assert(classfileManager.generatedClasses == (if (forked) Set() else Set(classfile)))
+    assert(
```
This is a hack 😛I presume you're doing it because in macOS temporary files get a public symlink to a private symlink and here you are getting the dealiased paths for the generated class files. Can we instead dealias `classfile`?

eed3si9n replied:
ok.

eed3si9n replied:
https://github.com/sbt/zinc/pull/640/commits/23485a8fa5ebc208805727c1121f9126874022dd

```scala
  def dealiasSymlinks(xs: HashSet[File]): HashSet[String] =
    xs map { x =>
      x.getCanonicalPath
    }
```

jvican commented on internal/zinc-compile-core/src/test/scala/sbt/internal/inc/javac/JavaCompilerSpec.scala:
```diff
@@ -71,7 +79,7 @@ class JavaCompilerSpec extends UnitSpec with DiagrammedAssertions {
     val (result, problems) = compile(compiler, Seq(knownSampleErrorFile), Seq("-deprecation"))
     assert(result == false)
-    assert(problems.size == 6)
+    assert(problems.size >= 5)
```
Can we have two different asserts depending on the JDK the tests run? If JDK8 or 9 go for 6 errors, otherwise go for 5.

eed3si9n replied:
ok.

eed3si9n replied:
https://github.com/sbt/zinc/pull/640/commits/23485a8fa5ebc208805727c1121f9126874022dd

```scala
    assert(problems.size == {
      sys.props("java.specification.version") match {
        case "1.8" | "9" => 6
        case _           => 5
      }
    })
```

jvican commented on internal/zinc-compile-core/src/test/scala/sbt/internal/inc/javac/JavaCompilerSpec.scala:
```diff
@@ -82,9 +90,8 @@ class JavaCompilerSpec extends UnitSpec with DiagrammedAssertions {
-    val (result, problems) =
+    val (_, problems) =
       doc(compiler, Seq(knownSampleErrorFile), Seq("-d", out.getAbsolutePath))
-    assert(result)
```
What's the motivation for this change?

eed3si9n replied:
`doc(...)` correctly returns `false` here so the assert fails. Maybe that's also a JDK dependent behavior?

jvican replied:
Maybe we can also show that in the test, as in the case of the error mismatch?

eed3si9n replied:
```scala
    // exit code for `javadoc` commandline is JDK dependent
    assert(result == {
      sys.props("java.specification.version") match {
        case "1.8" | "9" => true
        case _           => false
      }
    })
```

jvican commented on zinc/src/sbt-test/source-dependencies/ext/A.scala:
```diff
@@ -1,5 +0,0 @@
-import sun.net.spi.nameservice.dns.DNSNameService
```
Can we instead move this to `pending`?

eed3si9n replied:
If it passes in JDK 8 but fails on JDK 11, then we can't.

jvican replied:
sounds good

jvican reviewed (changes requested):
LGTM, happy to see the tests passing in JDK11 :+1: I reviewed https://github.com/sbt/zinc/pull/640/commits/a4b98ce596825378fb4d137d41f9bc320c000c68 in the other PR. The comment in this PR concern the rest of the commits.

eed3si9n commented:
@jvican Rebased to develop, and addressed the review comments.

jvican reviewed (approved):
This is great work @eed3si9n 💯 Thanks for looking into this before I did :smile:
